### DISCUSSION TITLE
User-created Libraries for Temperaments, Scales, Code/Music Samples, Sound Samples, and Image Samples

### DISCUSSION THREAD

**Discussion by pikurasa** [sugarlabs/musicblocks-v4#45] [Ideas]
Right now users can publish projects publicly to the planet. This is very nice, of course.

However, sometimes users may want to publish/share/remix/use something more foundational to/from the community. In some ways it is akin to creating/sharing/remixing libraries.

I think we would want the following libraries:
* Custom Temperaments | Users share their custom temperaments | Blocks: Define Temperament; Custom Pitch: Custom Temperament (from the pie menu, calling to define temperament)
* Custom Scales | Users share their custom scales/modes | Blocks: Define Scale: Custom Mode; Custom Mode (from the pie menu, calling to define mod)
* Custom Code/Music Samples | Users share action blocks to the planet | Blocks: Action Blocks (Q: Do we want to allow publication of Start Blocks for this?) <-- Like "Save Action Block" feature, but published to a server for sharing.
* Sound Samples | Users share their (recorded) custom sounds | Blocks: Sample Block
* Image Samples | Users share their images | Blocks: Image block

For temperament and scale/mode, I imagine we can provide some mathematical statistics for published libraries, and that users can search the libraries based on indexed parameters. For example "Number of pitches" could be an important statistic for both temperament and mode.

For Custom Code/Music the intention would be to publish more focused code. The intention, for the music side, would be a place where users could quickly find some musical snippets to pull into their current projects.

For sound samples and image samples, as discussed before, we would need some way to moderate what is published. That being said, of all five of these categories, I put these at the bottom of the list in terms of their importance for learning core concepts (although I know they would be loads of fun for kids).

**Comment by walterbender** [comment]
Is there any current limitation to using the "merge" project for these types of things?

**Reply by pikurasa** [reply]
Are you suggesting that users publish these sorts of things to our current planet?

**Comment by walterbender** [comment]
Maybe we can address this by adding some new tag types

**Reply by pikurasa** [reply]
Fine by me. If there are any design choices that could be made to better invite the user to contribute to these categories, I think that would be great.

The one thing I really like about the "My Blocks" feature, which would be nice for the planet is that those saved Action blocks are shorter snippets, rather than a whole project, which includes the Start Blocks and everything else.

**Comment by walterbender** [comment]
Maybe we should roll back the requirement to have a start block?

**Comment by walterbender** [comment]
As per our discussion:

(1) Devin will identify addition blocks that might be added to MyBlocks (at present is is only Action Blocks)
(2) Walter will remove the start block requirement for a project (the run button will do nothing if there is no start block)
(3) Walter will add a new Planet Tag for "Code Snippets"

**Reply by pikurasa** [reply]
Additional Blocks:
1) Define Mode (which, I do not think is currently affecting "custom mode" atm. It is a bug.)
2) Define Temperament (which is working fine to affect "cutom temperament" -- hurrah!)
3) Sound Samples Blocks (and, I suppose the "set instrument" block that it is typically connected to?)
4) Set Image Blocks (i.e. "Media Block")

I think everything else is handled by being able to add Action Blocks to MyBlocks, but will keep my eyes and ears out.

**Comment by walterbender** [comment]
Half-way there. You can now save custom modes, temperaments, samples, and images to My Blocks.

**Reply by pikurasa** [reply]
Great!

**Comment by walterbender** [comment]
You no longer need to ave a start block. I added a Code Snippet tag (I will need Japanese :P) but I also need to add it to the Planet server tag list before it will appear. But meanwhile, you can save snippets to the Planet -- perhaps use "code snippet" in the description. This can include custom temperaments, modes, sound samples, avatars, etc.

**Reply by pikurasa** [reply]
The best I can think of is:
みじかいコード
短いコード
